src/Analyzers/CSharp/Analyzers/UsePatternCombinators/CSharpUsePatternCombinatorsAnalyzer.cs

```diff
@@ -0,0 +1,169 @@
+
+                case IBinaryOperation { OperatorKind: NotEquals } op:
+                    return Not.Create(ParseConstantPattern(op));
+
+                case IBinaryOperation { OperatorKind: ConditionalOr, Syntax: BinaryExpressionSyntax syntax } op:
+                    return ParseBinaryPattern(op, isDisjunctive: true, syntax.OperatorToken);
+
+                case IBinaryOperation { OperatorKind: ConditionalAnd, Syntax: BinaryExpressionSyntax syntax } op:
+                    return ParseBinaryPattern(op, isDisjunctive: false, syntax.OperatorToken);
+
+                case IBinaryOperation op when IsRelationalOperator(op.OperatorKind):
+                    return ParseRelationalPattern(op);
+
+                case IUnaryOperation { OperatorKind: UnaryOperatorKind.Not } op:
+                    return Not.Create(ParsePattern(op.Operand));
+
+                case IIsTypeOperation { Syntax: BinaryExpressionSyntax { Right: TypeSyntax type } } op:
+                    return new Type(type, op.ValueOperand);
+
+                case IIsPatternOperation { Pattern: { Syntax: PatternSyntax pattern } } op:
+                    return new Source(pattern, op.Value);
+
+                case IParenthesizedOperation op:
+                    return ParsePattern(op.Operand);
+            }
+
+            return null;
+        }
+
+        private static AnalyzedPattern? ParseBinaryPattern(IBinaryOperation op, bool isDisjunctive, SyntaxToken token)
+        {
+            var leftPattern = ParsePattern(op.LeftOperand);
+            if (leftPattern is null)
+                return null;
+
+            var rightPattern = ParsePattern(op.RightOperand);
+            if (rightPattern is null)
+                return null;
+
+            return Binary.Create(leftPattern, rightPattern, isDisjunctive, token);
+        }
+
+        private static ConstantResult DetermineConstant(IBinaryOperation op)
+        {
+            return (op.LeftOperand, op.RightOperand) switch
+            {
+                var (_, v) when IsConstant(v) => ConstantResult.Right,
+                var (v, _) when IsConstant(v) => ConstantResult.Left,
+                _ => ConstantResult.None,
+            };
+        }
+
+        private static AnalyzedPattern? ParseRelationalPattern(IBinaryOperation op)
+        {
+            return DetermineConstant(op) switch
+            {
+                ConstantResult.Left when op.LeftOperand.Syntax is ExpressionSyntax left
+                    // We need to flip the operator if the constant is on the left-hand-side.
+                    // This is because relational patterns only come in the prefix form.
+                    // For instance: `123 > x` would be rewritten as `x is < 123`.
+                    => new Relational(Flip(op.OperatorKind), left, op.RightOperand),
+                ConstantResult.Right when op.RightOperand.Syntax is ExpressionSyntax right
+                    => new Relational(op.OperatorKind, right, op.LeftOperand),
```

Comment:
it it just me or is it weird that we have a relational and the constant is on the right.  and then we pass 'right' in as the parameter before left.  something about that breaks my brain.  i.e. if i had  `x < 0, i would expect to do `new Relational(op, op.Left, op.Rigth)`.  i would be flipping the order of these parameters if the constant were on the left.

Reply:
technically a pattern match (target is pattern) should be represented as a separate node, but I didn't get into that for this feature just yet. I plan to do that as part of https://github.com/dotnet/roslyn/pull/32386 which this feature will reuse.

Reply:
sounds good.